### PR TITLE
enabling Cilium CNI with KubeEdge trial.

### PR DESCRIPTION
**Pull request by fujitatomoya**
Temporary PR to see the difference.

### CODE REVIEW

File: cloud/pkg/cloudhub/dispatcher/message_dispatcher.go [hunk 1]
```diff
@@ -265,6 +265,17 @@ func (md *messageDispatcher) enqueueAckMessage(nodeID string, msg *beehivemodel.
 		return
 	}
 
+	klog.Errorf("MessageRoute.Resource: %s", msg.GetResource())
```
Comment (fujitatomoya) [inline cloud/pkg/cloudhub/dispatcher/message_dispatcher.go:268]: these are just debug purpose messages? `klog.Errorf`

File: cloud/pkg/cloudhub/dispatcher/message_dispatcher.go [hunk 2]
```diff
@@ -265,6 +265,17 @@ func (md *messageDispatcher) enqueueAckMessage(nodeID string, msg *beehivemodel.
...
+	resourceType, _ := messagelayer.GetResourceType(*msg)
+	if resourceType == beehivemodel.ResourceTypeNamespace {
+		resourceNamespace = resourceName
+	}
+	klog.Errorf("resourceNamespace original: %s", resourceNamespace)
+	if resourceType != beehivemodel.ResourceTypeNode && resourceNamespace == "null" {
+		resourceNamespace = "default"
```
Comment (fujitatomoya) [inline cloud/pkg/cloudhub/dispatcher/message_dispatcher.go:275]: I think this does fall back to `default` namespace if that is not set in the message to downstream?

this is follow-up bugfix to address https://github.com/kubeedge/kubeedge/issues/4819#issuecomment-1628392110?

File: cloud/pkg/cloudhub/session/node_session.go [hunk 3]
```diff
@@ -386,6 +386,16 @@ func (ns *NodeSession) saveSuccessPoint(msg *beehivemodel.Message) {
 			return
 		}
 
+		klog.Errorf("MessageRoute.Resource: %s", msg.GetResource())
```
Comment (fujitatomoya) [inline cloud/pkg/cloudhub/session/node_session.go:389]: i believe this one is almost the same with `message_dispatcher`.

File: cloud/pkg/common/messagelayer/util.go [hunk 4]
```diff
@@ -62,6 +62,9 @@ func BuildResource(nodeID, namespace, resourceType, resourceID string) (resource
 func getElementByIndex(msg model.Message, index int) string {
 	sli := strings.Split(msg.GetResource(), constants.ResourceSep)
 	if len(sli) <= index {
+		if index == ResourceResourceTypeIndex {
+			return "node"
+		}
```
Comment (fujitatomoya) [inline cloud/pkg/common/messagelayer/util.go:67]: How come we can say this? we are not sure what is missing in the resource path? Do we have issue for this?

File: edge/pkg/metamanager/metaserver/handlerfactory/handler.go [hunk 5]
```diff
@@ -49,6 +49,38 @@ func NewFactory() *Factory {
...
 }
 
+func (f *Factory) Version() http.Handler {
+	if h, ok := f.getHandler("version"); ok {
+		return h
+	}
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	h := http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		result := []byte("{\"major\":\"1\", \"minor\":\"22\", \"gitVersion\":\"v1.22.17\", \"gitCommit\":\"a7736eaf34d823d7652415337ac0ad06db9167fc\", \"gitTreeState\":\"clean\", \"buildDate\":\"2022-12-08T11:42:04Z\", \"goVersion\":\"go1.16.15\", \"compiler\":\"gc\", \"platform\":\"linux/amd64\"}")
```
Comment (fujitatomoya) [inline edge/pkg/metamanager/metaserver/handlerfactory/handler.go:59]: the question is should we always reply k8s version against this request, right? that said, when starting metaserver, it should proxy the request to the api-server to ask this information at least once? after all, this is just a temporary patch to return the static version as it is used for our test.

File: edge/pkg/metamanager/metaserver/handlerfactory/handler.go [hunk 6]
```diff
@@ -49,6 +49,38 @@ func NewFactory() *Factory {
...
+}
+
+func (f *Factory) Healthz() http.Handler {
+	if h, ok := f.getHandler("healthz"); ok {
+		return h
+	}
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	h := http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		result := []byte("ok")
```
Comment (fujitatomoya) [inline edge/pkg/metamanager/metaserver/handlerfactory/handler.go:75]: same here, it just passes the request and return ok.

File: cloud/pkg/cloudhub/dispatcher/message_dispatcher.go (hunk 1 above)
Comment (TE-7000026184) [reply cloud/pkg/cloudhub/dispatcher/message_dispatcher.go:268]: Yes, just for temporary debugging.

File: cloud/pkg/cloudhub/dispatcher/message_dispatcher.go (hunk 2 above)
Comment (TE-7000026184) [reply cloud/pkg/cloudhub/dispatcher/message_dispatcher.go:275]: There are two different problems included in issue 4819.

Issue 4819 was firstly created for the problem when a namespace resource is requested. For example, handling "GET /api/v1/namespaces/kube-system".

The change here is for the problem when getting a resource within default namespace.  
`node/pc-k8s-192-168-0-121/null/namespace/kube-system`, `err="namespaces \"null\" not found"`.  
The `message_dispatcher` thinks `null` is the namespace name instead of in default namespace.

File: cloud/pkg/common/messagelayer/util.go (hunk 4 above)
Comment (TE-7000026184) [reply cloud/pkg/common/messagelayer/util.go:67]: I also don't think this is an elegant change.

The format of the message is `node/<node name>/<namespace name>/<resource type>/<resource name>`. Where first element "node" is a hard code string, and node name is added by edge side automatically. So there are at least two elements.  
If `len(sli) == 3`, the request message makes no sense, because when a normal application wants to get node, it does not specify namespace, and it is not needed by API server for providing node information.  

In theory, if following this format, the message for getting a node should be like "node/pc-k8s-192-168-0-121/null/node/pc-k8s-192-168-0-121".  
However, what I see in log is "node/pc-k8s-192-168-0-121".  
I think it's also a problem related to "null" namespace handling. I guess a "null" namespace cut this message string at the third element.

File: edge/pkg/metamanager/metaserver/handlerfactory/handler.go (hunk 5 above)
Comment (TE-7000026184) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:59]: According to Shelly, version API would be supported in v1.15 https://github.com/kubeedge/kubeedge/issues/4844#issuecomment-1653498983.

Just for satisfying cilium-agent temporarily here. It is not a frequently requested resource.

File: edge/pkg/metamanager/metaserver/handlerfactory/handler.go (hunk 6 above)
Comment (TE-7000026184) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:75]: Since the message is received and replied, and Edge autonomy is expected to be supported, I think it is okay to just reply "ok".

File: cloud/pkg/cloudhub/session/node_session.go (hunk 3 above)
Comment (TE-7000026184) [reply cloud/pkg/cloudhub/session/node_session.go:389]: Yes, codes are similar.

File: cloud/pkg/cloudhub/dispatcher/message_dispatcher.go (hunk 2 above)
Comment (fujitatomoya) [reply cloud/pkg/cloudhub/dispatcher/message_dispatcher.go:275]: that sounds promising to me, can you create the new issue to KubeEdge for this as follow-up of https://github.com/kubeedge/kubeedge/issues/4819?

File: cloud/pkg/common/messagelayer/util.go (hunk 4 above)
Comment (fujitatomoya) [reply cloud/pkg/common/messagelayer/util.go:67]: I am sorry i am not following your explanation. can you rephrase this a bit?

File: edge/pkg/metamanager/metaserver/handlerfactory/handler.go (hunk 5 above)
Comment (fujitatomoya) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:59]: okay, lets say this is open issue with https://github.com/kubeedge/kubeedge/issues/4844#issuecomment-1653498983

File: edge/pkg/metamanager/metaserver/handlerfactory/handler.go (hunk 6 above)
Comment (fujitatomoya) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:75]: I think this can be dedicated issue to address? registering `healthz` handler on edged?

**Comment by fujitatomoya** [comment]
@Windrow14 i am catching up, but i need more time to understand the implementation.

File: edge/pkg/metamanager/metaserver/handlerfactory/handler.go (hunk 5 above)
Comment (TE-7000026184) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:59]: https://github.com/kubeedge/kubeedge/pull/4959  
Someone is working on this.

File: cloud/pkg/common/messagelayer/util.go (hunk 4 above)
Comment (TE-7000026184) [reply cloud/pkg/common/messagelayer/util.go:67]: I see this ticket for "Object 'Kind' is missing" issue and a possible solution in its comment: https://github.com/kubeedge/kubeedge/issues/2956#issuecomment-1437702572.

So, I was trying to find out which messages are for getting node but without specifying the "Kind" by checking `len(sli)`.

File: cloud/pkg/cloudhub/dispatcher/message_dispatcher.go (hunk 2 above)
Comment (TE-7000026184) [reply cloud/pkg/cloudhub/dispatcher/message_dispatcher.go:275]: https://github.com/kubeedge/kubeedge/issues/5042 I'd like to discuss the solution with KubeEdge teams first, I don't have much confidence with my current patch...

Comment (fujitatomoya) [reply cloud/pkg/cloudhub/dispatcher/message_dispatcher.go:275]: Okay this requires updating KubeEdge v1.14 or later, and that means we need to use Kubernetes v1.24, that deprecated docker-shim. as open source perspective, this is resolved, i will go ahead to close this comment.

File: cloud/pkg/common/messagelayer/util.go (hunk 4 above)
Comment (fujitatomoya) [reply cloud/pkg/common/messagelayer/util.go:67]: I am not really sure the relationship between this specific work-around and https://github.com/kubeedge/kubeedge/issues/2956... could you please elaborate a bit? or you are not sure about this either?

Comment (fujitatomoya) [reply cloud/pkg/common/messagelayer/util.go:67]: > If len(sli) == 3, the request message makes no sense, because when a normal application wants to get node, it does not specify namespace, and it is not needed by API server for providing node information.

i see this does not make sense though.

File: edge/pkg/metamanager/metaserver/handlerfactory/handler.go (hunk 5 above)
Comment (fujitatomoya) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:59]: okay, i see the update. resolving this comment.

File: edge/pkg/metamanager/metaserver/handlerfactory/handler.go (hunk 6 above)
Comment (fujitatomoya) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go:75]: This is alive, this is required to have for enabling cilium CNI, right? could you add the description for this?

**Review by fujitatomoya — COMMENTED** [review]
@Windrow14 there are 2 left unresolved comments, can you check?
